flink-connectors/flink-connector-base/src/main/java/org/apache/flink/connector/base/source/hybrid/HybridSource.java

```diff
@@ -0,0 +1,216 @@
+
+    public HybridSource(SourceChain<T, ? extends SourceSplit, ?> sourceChain) {
+        Preconditions.checkArgument(!sourceChain.sources.isEmpty());
+        for (int i = 0; i < sourceChain.sources.size() - 1; i++) {
+            Preconditions.checkArgument(
+                    Boundedness.BOUNDED.equals(sourceChain.sources.get(i).f0.getBoundedness()),
+                    "All sources except the final source need to be bounded.");
+        }
+        this.sourceChain = sourceChain;
+    }
+
+    @Override
+    public Boundedness getBoundedness() {
+        return sourceChain.sources.get(sourceChain.sources.size() - 1).f0.getBoundedness();
+    }
+
+    @Override
+    public SourceReader<T, HybridSourceSplit> createReader(SourceReaderContext readerContext)
+            throws Exception {
+        List<SourceReader<?, ? extends SourceSplit>> readers = new ArrayList<>();
+        for (Tuple2<Source<T, ? extends SourceSplit, ?>, ?> source : sourceChain.sources) {
+            readers.add(source.f0.createReader(readerContext));
+        }
+        return new HybridSourceReader(readerContext, readers);
+    }
+
+    @Override
+    public SplitEnumerator<HybridSourceSplit, HybridSourceEnumState> createEnumerator(
+            SplitEnumeratorContext<HybridSourceSplit> enumContext) {
+        return new HybridSourceSplitEnumerator(enumContext, sourceChain);
+    }
+
+    @Override
+    public SplitEnumerator<HybridSourceSplit, HybridSourceEnumState> restoreEnumerator(
+            SplitEnumeratorContext<HybridSourceSplit> enumContext, HybridSourceEnumState checkpoint)
+            throws Exception {
+        return new HybridSourceSplitEnumerator(
+                enumContext, sourceChain, checkpoint.getCurrentSourceIndex());
+    }
+
+    @Override
+    public SimpleVersionedSerializer<HybridSourceSplit> getSplitSerializer() {
+        List<SimpleVersionedSerializer<SourceSplit>> serializers = new ArrayList<>();
+        sourceChain.sources.forEach(
+                t -> serializers.add(castSerializer(t.f0.getSplitSerializer())));
+        return new SplitSerializerWrapper<>(serializers);
+    }
+
+    @Override
+    public SimpleVersionedSerializer<HybridSourceEnumState> getEnumeratorCheckpointSerializer() {
+        List<SimpleVersionedSerializer<Object>> serializers = new ArrayList<>();
+        sourceChain.sources.forEach(
+                t -> serializers.add(castSerializer(t.f0.getEnumeratorCheckpointSerializer())));
+        return new HybridSourceEnumStateSerializer(serializers);
+    }
+
+    private static <T> SimpleVersionedSerializer<T> castSerializer(
+            SimpleVersionedSerializer<? extends T> s) {
+        @SuppressWarnings("rawtypes")
+        SimpleVersionedSerializer s1 = s;
+        return s1;
+    }
+
+    /** Serializes splits by delegating to the source-indexed split serializer. */
+    public static class SplitSerializerWrapper<SplitT extends SourceSplit>
```

Comment:
Should we move this nested class out as an independent class to keep the HybridSource class a little shorter? that would also match the style of other source impls